stdlib/public/core/UTF16.swift

```diff
@@ -434,3 +434,253 @@ extension Unicode.UTF16.ForwardParser: Unicode.Parser, _UTFParser {
+  input: inout UnsafePointer<UInt16>,
+  inputEnd: UnsafePointer<UInt16>,
+  output: inout UnsafeMutablePointer<UInt8>,
+  outputEnd: UnsafePointer<UInt8>,
+  repairing: Bool
+) -> Bool {
+  for _ in 0 ..< 16 {
+    switch processScalarFallback(
+      input: &input,
+      inputEnd: inputEnd,
+      output: &output,
+      outputEnd: outputEnd,
+      repairing: repairing
+    ) {
+    case .invalid:
+      return false
+    case .multiByte:
+      return true //found the non-ASCII, try starting a new SIMD batch
+    case .singleByte:
+      continue
+    }
+  }
+  Builtin.unreachable()
+}
+
+internal func transcodeUTF16ToUTF8(
+  UTF16CodeUnits: UnsafeBufferPointer<Unicode.UTF16.CodeUnit>,
+  into outputBuffer: UnsafeMutableBufferPointer<Unicode.UTF8.CodeUnit>,
+  repairing: Bool = true
+) -> Int {
+  let inCount = UTF16CodeUnits.count
+  let outCount = outputBuffer.count
+  guard inCount > 0, outCount > 0 else { return 0 }
+  var input = UTF16CodeUnits.baseAddress.unsafelyUnwrapped
+  let inputEnd = input + inCount
+  let inputEnd256 = input + inCount & ~(MemoryLayout<SIMD16<UInt16>>.stride &- 1)
+  var output = outputBuffer.baseAddress.unsafelyUnwrapped
+  let outputStart = output
+  let outputEnd = output + outCount
+  let outputEnd256 = output + outCount & ~(MemoryLayout<SIMD16<UInt16>>.stride &- 1)
+  
+  while input < inputEnd256 && output < outputEnd256 {
+    if let asciiBlock = allASCIIBlock(at: input) {
+      // All ASCII: transcode directly
+      for i in 0..<16 {
+        (output + i).initialize(to: asciiBlock[i])
+      }
+      input += 16
+      output += 16
+    } else {
+      if !processNonASCIIChunk(
+        input: &input,
+        inputEnd: inputEnd,
+        output: &output,
+        outputEnd: outputEnd,
+        repairing: repairing
+      ) {
+        return output - outputStart
+      }
+    }
+  }
+  // Finish any remaining code units using fallback scalar loop
+  while input < inputEnd && output < outputEnd {
+    let scalarFallBackResult = processScalarFallback(
+      input: &input,
+      inputEnd: inputEnd,
+      output: &output,
+      outputEnd: outputEnd,
+      repairing: repairing
+    )
+    if scalarFallBackResult ~= .invalid {
+      return output - outputStart
+    }
+  }
+  return output - outputStart
+}
+
+internal func utf8Length(
+  of UTF16CodeUnits: UnsafeBufferPointer<Unicode.UTF16.CodeUnit>,
+  repairing: Bool = true
+) -> (Int, isASCII: Bool) ? {
+  let inCount = UTF16CodeUnits.count
+  guard inCount > 0 else { return 0 }
+  var input = UTF16CodeUnits.baseAddress.unsafelyUnwrapped
+  let inputEnd = input + inCount
+  let inputEnd256 = input + inCount & ~(MemoryLayout<SIMD16<UInt16>>.stride &- 1)
+  var count = 0
+  var isASCII = true
+  while input < inputEnd256 {
+    if let asciiBlock = allASCIIBlock(at: input) {
+      input += 16
+      count += 16
+    } else {
+      isASCII = false
+      var tmp: (
+        UInt32, UInt32, UInt32, UInt32, UInt32, UInt32, UInt32, UInt32
```

Comment:
Making a temporary buffer here is sort of awful and I want to improve it at some point, but it's also not really hurting anything and simplifies the rest of the code a lot